wasm/wat/token.go

```diff
@@ -0,0 +1,217 @@
+
+	// tokenId is a sequence of idChar characters prefixed by a '$':
+	//
+	// For example, in the below, '$y' is an id:
+	//		local.get $y
+	//		i32.const 6
+	//		i32.lt_s
+	//
+	// See https://www.w3.org/TR/wasm-core-1/#text-id
+	tokenId
+
+	// tokenLParen is a left paren: '('
+	tokenLParen
+
+	// tokenRParen is a right paren: ')'
+	tokenRParen
+
+	// tokenReserved is a sequence of idChar characters which are neither a tokenId nor a tokenString.
+	//
+	// For example, '0$y' is a tokenReserved, because it doesn't start with a letter or '$'.
+	//
+	// See https://www.w3.org/TR/wasm-core-1/#text-reserved
+	tokenReserved
+)
+
+// tokenNames is index-coordinated with tokenType
+var tokenNames = []string{
+	"tokenInvalid",
+	"tokenKeyword",
+	"tokenUN",
+	"tokenSN",
+	"tokenFN",
+	"tokenString",
+	"tokenId",
+	"tokenLParen",
+	"tokenRParen",
+	"tokenReserved",
+}
+
+// String returns the string name of this token.
+func (t tokenType) String() string {
+	return tokenNames[t]
+}
+
+// constants below help format a somewhat readable lookup table that eases identification of tokens.
+const (
+	// xx is an invalid token start byte
+	xx = tokenInvalid
+	// xs is the start of tokenString ('"')
+	xs = tokenString
+	// xi is the start of tokenId ('$')
+	xi = tokenId
+	// lp is the start of tokenLParen ('(')
+	lp = tokenLParen
+	// rp is the start of tokenRParen (')')
+	rp = tokenRParen
+	// un is the start of a tokenUN (or tokenFN)
+	un = tokenUN
+	// sn is the start of a tokenSN (or tokenFN)
+	sn = tokenSN
+	// xk is the start of a tokenKeyword
+	xk = tokenKeyword
+	// xr is the start of tokenReserved (a valid, but not defined above).
+	xr = tokenReserved
+)
+
+// firstTokenByte is information about the firstTokenByte byte in a token. All expected token starts are ASCII, but we
+// switch to avoid a range check.
+var firstTokenByte = [256]tokenType{
+	//   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0x00-0x0F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0x10-0x1F
+	xx, xr, xs, xr, xi, xr, xr, xr, lp, rp, xr, sn, xx, sn, xr, xr, // 0x20-0x2F
+	un, un, un, un, un, un, un, un, un, un, xr, xx, xr, xr, xr, xr, // 0x30-0x3F
+	xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, // 0x40-0x4F
+	xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xr, xx, xr, xx, xr, xr, // 0x50-0x5F
+	xr, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, // 0x60-0x6F
+	xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xk, xx, xr, xx, xr, xx, // 0x70-0x7F
+}
+
+// idChar is a printable ASCII character that does not contain a space, quotation mark, comma, semicolon, or bracket.
+// See https://www.w3.org/TR/wasm-core-1/#text-idchar
+var idChar = buildIdChars()
+
+func buildIdChars() (result [256]bool) {
+	for i := 0; i < 128; i++ {
+		result[i] = _idChar(byte(i))
+	}
+	return
+}
+
+func _idChar(ch byte) bool {
+	switch ch {
+	case '!':
+		fallthrough
+	case '#':
+		fallthrough
```

Comment:
I think we could reduce the # of lines like this:
```suggestion
	case '!', '#', .... :
		return true
```

Reply:
ah right I'll do that.